### ISSUE TITLE
wagtailimages squashed migration breaks custom image migrations from Wagtail 1.7

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#5067]
### Issue Summary

(Unconfirmed, reported on the Slack support channel)

If a Wagtail project has a custom image model, and it has existed since Wagtail 1.7 or earlier, the app containing the custom image model will have migrations referencing the `Filter` model, such as the data migration detailed in https://docs.wagtail.io/en/v2.4/releases/1.7.html#custom-image-models-require-a-data-migration-for-the-new-filter-spec-field . Running these migrations (when bringing up a new database from scratch, or running a test suite) fails on Wagtail 2.4 with `ValueError: Related model 'wagtailimages.Filter' cannot be resolved`; presumably the presence of the squashed migration in wagtailimages (#4917) means that the migrations for the custom model only get run after wagtailimages migration 0021, and at that point the `Filter` model does not exist in the frozen ORM.

Related (the custom data migration is itself not squashable with `squashmigrations` due to the presence of local function references): #3259

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: no

### Technical details

* Python version: 3.5
* Django version: 2.1
* Wagtail version: 2.4

```Traceback (most recent call last):
  File "manage.py", line 25, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/test.py", line 26, in run_from_argv
    super().run_from_argv(argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 316, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 353, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/test.py", line 56, in handle
    failures = test_runner.run_tests(test_labels)
  File "/usr/local/lib/python3.5/site-packages/django_nose/runner.py", line 308, in run_tests
    result = self.run_suite(nose_argv)
  File "/usr/local/lib/python3.5/site-packages/django_nose/runner.py", line 245, in run_suite
    addplugins=plugins_to_add)
  File "/usr/local/lib/python3.5/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/local/lib/python3.5/unittest/main.py", line 95, in __init__
    self.runTests()
  File "/usr/local/lib/python3.5/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/usr/local/lib/python3.5/site-packages/nose/core.py", line 50, in run
    wrapper = self.config.plugins.prepareTest(test)
  File "/usr/local/lib/python3.5/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/usr/local/lib/python3.5/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/usr/local/lib/python3.5/site-packages/django_nose/plugin.py", line 82, in prepareTest
    self.old_names = self.runner.setup_databases()
  File "/usr/local/lib/python3.5/site-packages/django_nose/runner.py", line 495, in setup_databases
    return super(NoseTestSuiteRunner, self).setup_databases()
  File "/usr/local/lib/python3.5/site-packages/django/test/runner.py", line 551, in setup_databases
    self.parallel, **kwargs
  File "/usr/local/lib/python3.5/site-packages/django/test/utils.py", line 174, in setup_databases
    serialize=connection.settings_dict.get('TEST', {}).get('SERIALIZE', True),
  File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/creation.py", line 68, in create_test_db
    run_syncdb=True,
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 148, in call_command
    return command.execute(*args, **defaults)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 353, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/migrate.py", line 203, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/local/lib/python3.5/site-packages/django/db/migrations/operations/models.py", line 91, in database_forwards
    schema_editor.create_model(model)
  File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/schema.py", line 263, in create_model
    definition, extra_params = self.column_sql(model, field)
  File "/usr/local/lib/python3.5/site-packages/django/db/backends/base/schema.py", line 157, in column_sql
    db_params = field.db_parameters(connection=self.connection)
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 966, in db_parameters
    return {"type": self.db_type(connection), "check": self.db_check(connection)}
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 963, in db_type
    return self.target_field.rel_db_type(connection=connection)
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 878, in target_field
    return self.foreign_related_fields[0]
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 632, in foreign_related_fields
    return tuple(rhs_field for lhs_field, rhs_field in self.related_fields if rhs_field)
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 619, in related_fields
    self._related_fields = self.resolve_related_fields()
  File "/usr/local/lib/python3.5/site-packages/django/db/models/fields/related.py", line 604, in resolve_related_fields
    raise ValueError('Related model %r cannot be resolved' % self.remote_field.model)
ValueError: Related model 'wagtailimages.Filter' cannot be resolved
```

Custom data migration from the app which produced the above error:
```
# -*- coding: utf-8 -*-
# Generated by Django 1.10.1 on 2016-10-22 07:38
from __future__ import unicode_literals

from django.db import migrations
from wagtail.images.utils import get_fill_filter_spec_migrations


class Migration(migrations.Migration):

    dependencies = [
        ('wagtailextendedimages', '0005_auto_20161022_0738'),
    ]

    forward, reverse = get_fill_filter_spec_migrations('wagtailextendedimages', 'ExtendedRendition')
    operations = [
        migrations.RunPython(forward, reverse),
    ]
```

**Comment by rinti** [comment]
Hi

I'm the one who ran in to this. I also had those migrations referencing wagtailimages.Filter:

```
# -*- coding: utf-8 -*-
# Generated by Django 1.10.1 on 2016-10-14 16:09
from __future__ import unicode_literals

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import taggit.managers
import wagtail.core.models
import wagtail.images.models
import wagtail.search.index


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('wagtailcore', '0029_unicode_slugfield_dj19'),
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
        ('wagtailimages', '0013_make_rendition_upload_callable'),
        ('taggit', '0002_auto_20150616_2121'),
    ]

    operations = [
        migrations.CreateModel(
            name='ExtendedImage',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('title', models.CharField(max_length=255, verbose_name='title')),
                ('file', models.ImageField(height_field='height', upload_to=wagtail.images.models.get_upload_to, verbose_name='file', width_field='width')),
                ('width', models.IntegerField(editable=False, verbose_name='width')),
                ('height', models.IntegerField(editable=False, verbose_name='height')),
                ('created_at', models.DateTimeField(auto_now_add=True, db_index=True, verbose_name='created at')),
                ('focal_point_x', models.PositiveIntegerField(blank=True, null=True)),
                ('focal_point_y', models.PositiveIntegerField(blank=True, null=True)),
                ('focal_point_width', models.PositiveIntegerField(blank=True, null=True)),
                ('focal_point_height', models.PositiveIntegerField(blank=True, null=True)),
                ('file_size', models.PositiveIntegerField(editable=False, null=True)),
                ('collection', models.ForeignKey(default=wagtail.core.models.get_root_collection_id, on_delete=django.db.models.deletion.CASCADE, related_name='+', to='wagtailcore.Collection', verbose_name='collection')),
                ('tags', taggit.managers.TaggableManager(blank=True, help_text=None, through='taggit.TaggedItem', to='taggit.Tag', verbose_name='tags')),
                ('uploaded_by_user', models.ForeignKey(blank=True, editable=False, null=True, on_delete=django.db.models.deletion.SET_NULL, to=settings.AUTH_USER_MODEL, verbose_name='uploaded by user')),
            ],
            options={
                'abstract': False,
            },
            bases=(wagtail.search.index.Indexed, models.Model),
        ),
        migrations.CreateModel(
            name='ExtendedRendition',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('file', models.ImageField(height_field='height', upload_to=wagtail.images.models.get_rendition_upload_to, width_field='width')),
                ('width', models.IntegerField(editable=False)),
                ('height', models.IntegerField(editable=False)),
                ('focal_point_key', models.CharField(blank=True, default='', editable=False, max_length=255)),
                ('filter', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='+', to='wagtailimages.Filter')),
                ('image', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='renditions', to='wagtailextendedimages.ExtendedImage')),
            ],
        ),
        migrations.AlterUniqueTogether(
            name='extendedrendition',
            unique_together=set([('image', 'focal_point_key')]),
        ),
    ]
```

```
# -*- coding: utf-8 -*-
# Generated by Django 1.10.1 on 2016-10-22 07:38
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('wagtailextendedimages', '0004_auto_20161020_0738'),
    ]

    operations = [
        migrations.AddField(
            model_name='extendedrendition',
            name='filter_spec',
            field=models.CharField(blank=True, db_index=True, default='', max_length=255),
        ),
        migrations.AlterField(
            model_name='extendedrendition',
            name='filter',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='+', to='wagtailimages.Filter'),
        ),
    ]
```

```
# -*- coding: utf-8 -*-
# Generated by Django 1.10.5 on 2017-03-20 13:09
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('wagtailextendedimages', '0014_auto_20170117_1933'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='extendedrendition',
            name='filter',
        ),
        migrations.AlterField(
            model_name='extendedrendition',
            name='focal_point_key',
            field=models.CharField(blank=True, default='', editable=False, max_length=16),
        ),
    ]
```

**Comment by gasman** [comment]
Now reproduced independently, using this test project that replicates a Wagtail project with a custom image model having gone through upgrades from 1.7 to 2.5: https://github.com/gasman/wagtail-custom-image-upgrade

Essentially, the `replaces = [...]` line in Wagtail's `0001_squashed_0021` migration causes Django to forcibly ignore the presence of the original migrations 0001 ... 0021 (except when running on an installation that's midway through those migrations). This is bad news for [a custom image migration created on Wagtail 1.7](https://github.com/gasman/wagtail-custom-image-upgrade/blob/master/home/migrations/0003_customimage.py) - the presence of the foreign key to `Filter` means that it can only be run at a point after `wagtailimages.0001_initial` and before `wagtailimages.0019_delete_filter` - a point that no longer exists from Django's perspective. It doesn't look like Django can be coaxed into accepting it through a declaration like `run_before = [('wagtailimages', '0019_delete_filter')]` either...

So, I think the remaining route for working around this is going to be changing the ForeignKey to something harmless, like an `IntegerField`...

**Comment by gasman** [comment]
Yep - changing the line

    ('filter', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='+', to='wagtailimages.Filter')),

to

    ('filter', models.IntegerField(blank=True, null=True)),

allows the migration to complete on an up-to-date Wagtail version. I'm pretty confident this change is safe: an installation that has already run the migration will ignore the change, while an installation that hasn't run it will immediately follow it with the subsequent migration to remove the column (a migration which would have been created on the first run of `makemigrations` after upgrading to 1.9) and end up in the same state as an installation that followed the original migration.

I don't think there's a good way to fix this from Wagtail's side to make the original migration work again: in retrospect adding the squashed migration was probably a mistake (see https://github.com/wagtail/wagtail/issues/4965#issuecomment-503570537 for further thoughts on the subject...) but it can't be rolled back as that'll break any Wagtail installations that already have it in their migration history. Any solution would presumably have to be layered on top of the migrations that exist now, and that'll add more potential points of failure.

So, I think the resolution is to document the above workaround in the 2.4 release notes. Annoyingly any change to those docs won't show up for anyone _looking at the 2.4 version of the docs_ unless we either move away from readthedocs's git-tag-based versioning, or make a 2.4.1 release. But it's better than nothing, and even having the fix documented (and googleable) here on this ticket is a start :-)

**Comment by KalobTaulien** [comment]
This solution got me most of the way there. I ended up running into this message afterwards:

```
App 'wagtailimages' doesn't have a 'Filter' model.
```

If anyone else ends up getting the "Filter model does not exist" error, take a look in your migration files for `get_fill_filter_spec_migrations`. I was able to simply delete this from my migration file without having to think about backwards compatibility and that solved my issue, but someone may need to dive into `wagtail.images.utils.get_fill_filter_spec_migrations` and reconstruct some of it to maintain forwards and backwards database migrations.
